### PR TITLE
GDR contrib- Suballocators fixes

### PR DESCRIPTION
**Pull request by yanivbl6**
This patch includes two fixes, all in the gdr contribution
The 1st one changes the ordering of the calls to gdr memory manager initialization and grpc initialization.

This is necessary because gdr initialization may initiates the GPU device and call GetCPUAllocator (or GetGPUAllocator), which are asserted to never occur before the registration of Suballocators to ProcessState's singleton (which happens on gdr memory manager init)

The 2nd fix is to change the cuda host sub allocators to be registered on both Numas, 0 and 1.
This fixed an error on my servers, and I think it should be fine on different setups too but please comment if you think I missed something.

@byronyi @poxvoculi

### CODE REVIEW

**Comment by byronyi** [comment]
Thanks for the fix! I will check on our environment soon.

**Comment by yanivbl6** [comment]
Great! Note that I experienced an hang during initialization with master branch, but this also happened in vanilla grpc so it's probably unrelated. I didnt isolate the grpc problem yet but grpc and patched gdr worked with the 5 days old branch. (96a633367ecd5ae9b31e128c2436b1a3f81b27fd)

**Comment by drpngx** [comment]
@byronyi let me know when you have confirmed that it works.

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (poxvoculi) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: This seems fine for now.  NUMA specific allocation has not yet been turned on, pending bringing TF's eigen version up to date.   It might be to add a visitor for every available NUMA node by using port::NUMANumNodes.

**Review by poxvoculi — DISMISSED** [review]

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (yanivbl6) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: thanks for the tip. added it to latest commit.

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: This should be moved out of the above `if (IsGDRAvailable())`, as the host memory should be registered regardless. Otherwise, it fails for GDR incapable GPUs, such as those GTX cards.

**Review by byronyi — CHANGES_REQUESTED** [review]
Thanks for the fix!

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: Probably should change the name of `IsGDRAvailable` to `P2PKernelModuleEnabled` to avoid confusion.

Any suggestions?

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: Should rename `bus_id` to `numa_idx`, and move it inside the loop.

**Review by byronyi — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: @poxvoculi I saw different signatures for `AddGPUAllocVisitor(int bus_id, const SubAllocator::Visitor& visitor)` and `AddCUDAHostAllocVisitor(int numa_node, const SubAllocator::Visitor& visitor)`. It seems rather confusing.

Maybe change `bus_id` to `numa_node` for `AddGPUAllocVisitor` as well?

**Review by byronyi — CHANGES_REQUESTED** [review]
Please also remove [this line](https://github.com/tensorflow/tensorflow/blob/83e61f6060af9066e41dbac7dea44d860d66903a/tensorflow/contrib/gdr/gdr_memory_manager.cc#L269) and change `bus_id` [here](https://github.com/tensorflow/tensorflow/blob/83e61f6060af9066e41dbac7dea44d860d66903a/tensorflow/contrib/gdr/gdr_memory_manager.cc#L285-286) to `numa_node_`.

**Comment by byronyi** [comment]
With requested changes, I have tested the patch with CPU only, GPU without GDR, GPU with GDR but on different NUMA, and GPU with GDR on the same NUMA.

Thanks again!

**Comment by yanivbl6** [comment]
thanks for the comments. will go over them and re-commit tomorrow.

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (yanivbl6) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: GDR is the correct name here. Transporting between GPU's host pinned memory is technically just "GD".

I would like to eventually add the possibility to selectively avoid GPU-direct on slow PCie routes (Like QPI), but I prefer to just focus on hotfixes for now, till the upcoming changes regarding the transport contributions.

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: This line should also be moved above (i.e. outside `IsGDRAvailable()`)

**Comment by byronyi** [comment]
@drpngx LGTM

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (drpngx) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: Oh this is going to create a problem internally. Could you change that into a return error::InternalError?

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (drpngx) [inline tensorflow/contrib/gdr/gdr_memory_manager.cc]: Remove space. Consider using vlog.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/gdr/gdr_memory_manager.cc
Comment (byronyi) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: FYI: This only hits when turning VLOG>=2, and it is being used to debug memory corruption problems from hardware direct access. Like an assertion, and it’s easier to debug if we could keep the original stack instead of returning.

Thoughts?

Comment (drpngx) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: It's a policy that we have internally. You can LOG(ERROR) something and then return.

Comment (yanivbl6) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: sure, I will do it tomorrow morning.

Comment (yanivbl6) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: @drpngx  Will just re-committing the line back to the non-formatted version be sufficient to comply with the policy or do I need to force-push over the commit with that change?

Comment (byronyi) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: I guess the latter as they always squash the PR to a single commit :p

Comment (drpngx) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: Yes it doesn't matter how the commits are organized (I prefer a linear chain, just piling up so that we can match up the review history).

Comment (yanivbl6) [reply tensorflow/contrib/gdr/gdr_memory_manager.cc]: done

**Comment by drpngx** [comment]
@yanivbl6 there's some import issue with another patch (#22559). Could you pull rebase and push again?

**Comment by yanivbl6** [comment]
Ok, but the referenced patch was my initial starting point, and I have done some rebasing since without issues.

**Comment by drpngx** [comment]
@yifeif I'm not sure why it's not creating the CL after it's pulled and approved. Any idea?

**Comment by yanivbl6** [comment]
not sure. The checks from the force push aren't done yet, but I don't think that's the problem as I didn't see any collisions when re-basing.

@byronyi, I feel like I am delaying the fix with a code you already have for a while. Do you want me to close the PR so you can push what you have been testing?

I can try a new PR but other than that I am out of ideas.

**Comment by drpngx** [comment]
OK, the process got unstuck. Now testing internally and waiting for internal review.

**Review by poxvoculi — APPROVED** [review]

**Review by drpngx — APPROVED** [review]

**Comment by byronyi** [comment]
@yanivbl6 It is alright. GitHub went down yesterday so let’s just be patient to get this merged :)

**Pull request merged**
